Clean up exception handling in StashApiClient

proski opened the pull request:
The first commit adds WireMock based unit tests for the entire `StashApiClient.java`. In some cases, the tests indicate issues in the code under test. Those issues are documented.

The second commit changed exception handling in StashApiClient. Only specific checked exceptions are used, `RuntimeException` and `Exception` are banned. The callers are changes to handle the exceptions. That code was getting `RuntimeException` and ignored it. Now it's forces to deal with it properly.

proski commented:
Improved tests for loading PRs and PR comments from multiple pages to avoid some repetition.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClientTest.java:
```diff
@@ -20,4 +77,132 @@ public void testParsePullRequestMergeStatus() throws Exception {
+    int commentCounter = 0;
+    boolean isLastPage = false;
+
+    while (!isLastPage) {
```
I think this has to be extracted to a helper method, or we should instead write the data in a more direct ways

proski replied:
I don't envision more tests using multi-page data. We can make make those two cases share some code, but I'm not sure it would be an improvement. There are subtle differences between `getPullRequests` and `getPullRequestComments`, and the common code would need to take care of those differences. That would only add complexity.

I tried JSON in strings first, but it's hard to read without formatting.

I'm thinking of loading JSON from files.

jakub-bochenski replied:
I wasn't thinking about code reuse, just about moving the logic out of main test body to improve readability

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClientTest.java:
```diff
@@ -20,4 +77,132 @@ public void testParsePullRequestMergeStatus() throws Exception {
+  @Test
+  public void getPullRequestComments_gets_pull_requests_from_multiple_pages() throws Exception {
+    final int totalComments = 27;
+    final int pageSize = 10;
```
maybe use a smaller page size (e.g. 3), to make the test data smaller?

proski replied:
Yes, changing it to 8 comments on pages with 3 comments.

jakub-bochenski commented:
> I don't see a good way to embed JSON into Java code (https://openjdk.java.net/jeps/355 will hopefully change it), so I'm constructing JSON programmatically.

How about we use POJOs and then serialize them into Json with Jackson etc.? For simple objects (I don't expect anything else) the transformation should be obvious.

Another idea would be to go for spring-cloud-contract, too bad Stash doesn't publish it's API contracts (or does it?)

proski commented:
> How about we use POJOs and then serialize them into Json with Jackson etc.? For simple objects (I don't expect anything else) the transformation should be obvious.

I don't think is would be simpler than what we have now.

> Another idea would be to go for spring-cloud-contract, too bad Stash doesn't publish it's API contracts (or does it?)

A quick Google search suggests that they exist, but no indication where they are. Again, I don't expect it to be a simplification.

I'm going to test standalone JSON files first.

proski commented:
Added support for all public methods except `deleteRequest()`, which is public by mistake. Still need to test more exceptional conditions like connection reset. Need to test responses better, especially for POST.

It turns out WireMock has good support for configuring through JSON files, I'm going to try it.

proski commented:
Converted multi-page tests to JSON files.

Using dynamic port for HTTP. There default is port 8080, which can conflict with Jenkins.

Using `wiremock-jre8-standalone` in `pom.xml`, it's recommended if Java 8+ is required.

proski commented:
This PR has achieved 91.9% code coverage for `StashApiClient.java`, which is pretty impressive. Not only is the code under test executed, the outputs are checked.

One problem is that only "easy HTTPS" is being tested, but not HTTP or real HTTPS. I believe JUnit has a way to run tests with multiple parameters, I'll look into that. In any case, unit testing real HTTPS can be tricky.

The unit tests have discovered multiple issues. I documented them in the code.

* RuntimeException is thrown, but the callers may not be prepared for it, as it's not a checked exception
 * Many calls ignore HTTP errors, especially "malformed response"
 * deletePullRequestComment() gives no indication whether the call has succeeded
 * mergePullRequest() throws on 409 Conflict instead of returning false as apparently intended
 * There are no checks whether the HTTP code indicates an error for the specific request
 * POST requests throw on 204 No Content
 * Timeouts are not testable without very slow tests, as the timings are not configurable

Applying the first commit from #68 on top of this PR (i.e. upgrading the HTTP client) breaks several tests. On the positive side, some errors are not ignored anymore. On the negative side, `RuntimeException` is thrown in some cases where `StashApiException` was thrown.

I think the best course of action would be to completely remove `RuntimeException` from the code first. That will give us visibility into what exceptions we should handle.

proski committed:
StashApiClientTest: Add WireMock based tests for StashApiClient

Document discovered issues.

proski commented:
I've added a commit to clean up exception handling. It would be better to update the HTTP client if we have a complete picture which exceptions are thrown and what code is handling them.

I didn't try to use perfect messages in the exceptions, I only wanted then to be better than what it is now. I'm trying to keep this PR short so it can be reviewed and merged quickly.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildListener.java:
```diff
@@ -61,7 +64,17 @@ public void onCompleted(AbstractBuild<?, ?> abstractBuild, @Nonnull TaskListener
+              "%s: cannot delete Build Start comment for pull request %s: %s",
+              abstractBuild.getParent().getDisplayName(), cause.getPullRequestId(), e);
+      logger.info(logmsg);
+      listener.getLogger().println(logmsg);
```
I think logging to the build output is enough
I would only log to global logger if there is no build to log to

proski replied:
OK, fixed. Good point.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -110,7 +110,8 @@ public void init() {
 
-  public String postBuildStartCommentTo(StashPullRequestResponseValue pullRequest) {
+  public String postBuildStartCommentTo(StashPullRequestResponseValue pullRequest)
+      throws StashApiException {
```
do we need to add this to the signature?

proski replied:
Yes, it's needed, and it's a good thing. The compiler is checking for me that the exception is either handled or passed to the caller.

This is not the right place to handle the exception. The caller knows more about the context and can do more than just log. I believe the failure to post the initial comment should lead to skipping the build. If we cannot post comments, the builds would be run continuously. If it's a temporary issue, the build will be attempted again when the issue goes away.

This indirection could be removed eventually. The `StashApiClient` object should not be owned by `StashRepository`. It should be available through `StashBuildTrigger` or through `StashCause` to all objects that deal with Bitbucket Server.

jakub-bochenski replied:
OK, what got me confused is that we seem to be catching and handling some StashApiExceptions inside this code already?

proski replied:
We handle `StashApiException` in that file, but not in that method.
`postBuildStartCommentTo()` should really be `private`, it is only called by `addFutureBuildTasks()` in the same class. `addFutureBuildTasks()` does the handling because it can do something meaningful. It skips building the PR, which is the right thing to do. If the plugin cannot post a comment indicating that the build starts, chances are it won't be able to post a comment about the build result, and then the PR would be considered eligible for building again, wasting Jenkins server resources.
As it stands now (i.e. without this PR), `RuntimeException` thrown by `StashApiClient` propagates all the way to the Jenkins core. The iteration over open pull requests stops when the exception is thrown.
If we handled exceptions in `postBuildStartCommentTo()`, we would need a way to indicate failure by other means, perhaps by returning `null` as the comment ID. We went through that in #93, it was your idea to use an exception instead of returning `null`.

jakub-bochenski replied:
OK, thanks for the clarification.
Maybe it would be useful to add this explanation (or parts of) as comment/javadoc

proski replied:
I'm trying to code self-explanatory. The compiler doesn't leave much freedom which checked exceptions I list on the interface, especially when combined with FindBugs/SpotBugs. The problem with long comments is that they become liabilities. They need to be edited as the code evolves, or they get out of date.

jakub-bochenski replied:
Agree in general, though I think if you added the first paragraph next to the place where we catch the exception it could be easily maintained. The only change I expect is if we stop handling it there, but it's easy to remove the comment then.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -311,7 +322,8 @@ public void addFutureBuildTasks(Collection<StashPullRequestResponseValue> pullRe
 
-  public void deletePullRequestComment(String pullRequestId, String commentId) {
+  public void deletePullRequestComment(String pullRequestId, String commentId)
+      throws StashApiException {
```
do we need to add this to the signature?

proski replied:
Yes. The callers are in a better position to handle the exception. In particular, `StashBuildListener` can post the issue to the build output.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -355,14 +367,22 @@ public void postFinishedComment(
 
-  private boolean isPullRequestMergeable(StashPullRequestResponseValue pullRequest) {
+  private boolean isPullRequestMergeable(StashPullRequestResponseValue pullRequest)
+      throws StashApiException {
```
do we need to add this to the signature?

proski replied:
Yes, at least for now. The caller is already logging if `false` is returned. It's better to log from one place.

Eventually we should provide better logging about the reason why a particular PR is not being built, but it's outside the scope of this PR. I'm thinking we could add a page showing the polling status with all PRs seen and the reason why the plugin decided to build or not to build each of them e.g.

```
Last poll at 2019-06-30 9:00
8 pull requests in OPEN state
PR-120 build scheduled
PR-121 already built
PR-122 build queued
PR-123 build running
PR-124 "NO TEST" seen
PR-125 cannot be merged
PR-126 is in conflicted state
PR-127 error checking merge and conflict status
```

proski committed:
Clean up exception handling in StashApiClient

Stop using RuntimeException, it's an unchecked exception. The callers may
be unaware that they need to handle it.

Stop using Exception. Always use most specific exception types.

Always rethrow caught exceptions by wrapping them in StashApiClient. Let
the callers handle them, as they are in a better position to chose what
to do and emit meaningful logs.

Don't print stack trace in StashApiClient. Print it when the exception is
handled. Some callers log the stack trace to the build log, not to the
global Jenkins log.

Don't log error messages and throw StashApiException at the same time,
unless the log message contains information that is not added to the
exception.

Adjust "throws" declarations. Catch errors where appropriate and log
them.

Document functions that throw StashApiException. Add comments explaining
how StashApiException is handled.

Adjust unit tests for the new behavior.

proski commented:
It turns out I removed stack traces in many places, please don't merge this PR until I fix that issue.

proski commented:
Should be good to merge now.

proski commented:
I added comments everywhere `StashApiException` is handled. I also added javadoc comments to the functions that throw `StashApiException` in `StashRepository`.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildListener.java:
```diff
@@ -61,7 +67,19 @@ public void onCompleted(AbstractBuild<?, ?> abstractBuild, @Nonnull TaskListener
+
+    // Delete the "Build Started" comment, as it gets replaced with a comment
+    // about the build result. Failure to delete the comment is not fatal, it's
+    // reported to the build log.
```
This comment is OK, though I think we could do with an even shorter one. Something like `handle here, because it's not fatal` inside the catch block

proski replied:
I'll keep it in mind for the future. Let's just leave the comments as is.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -360,14 +402,41 @@ public void postFinishedComment(
+      this.client.postPullRequestComment(pullRequestId, comment);
+    } catch (StashApiException e) {
+      logger.log(
+          Level.INFO,
```
actually, since this will cause rebuild I think we could make it a warning

proski replied:
Good point. Done.

proski commented:
Added unit tests for `StashBuildListener`.

proski commented:
Added unit tests for the changes to `StashRepository.java`.

proski commented:
Added checks for authentication, we really don't want to break it.
